README.md
# WebSocketJSONStream

[![npm version](https://badge.fury.io/js/%40teamwork%2Fwebsocket-json-stream.svg)](https://badge.fury.io/js/%40teamwork%2Fwebsocket-json-stream)
[![Build Status](https://travis-ci.org/Teamwork/websocket-json-stream.svg?branch=master)](https://travis-ci.org/Teamwork/websocket-json-stream)
[![Coverage Status](https://coveralls.io/repos/github/Teamwork/websocket-json-stream/badge.svg)](https://coveralls.io/github/Teamwork/websocket-json-stream)

A nodejs stream wrapper for WebSocket connections.

## Usage

```js
const WebSocket = require('ws')
const WebSocketJSONStream = require('@teamwork/websocket-json-stream')

const stream = new WebSocketJSONStream(new WebSocket(url))
// ...

new WebSocket.Server({ server }).on('connection', ws => {
    const stream = new WebSocketJSONStream(ws)
    // ...
})
```

See [example.js](./example.js) for a working usage example.

## Closing a WebSocket via its stream

Calling [`stream.end()`](https://nodejs.org/api/stream.html#stream_writable_end_chunk_encoding_callback) or [`stream.destroy()`](https://nodejs.org/api/stream.html#stream_writable_destroy_error) will close the WebSocket connection.

When a WebSocket is closed either by the server or the client, a [`CloseEvent`](https://developer.mozilla.org/en-US/docs/Web/API/CloseEvent) will be emitted. CloseEvents have both a numeric `code` and a string `reason` property that may be used to indicate the type of closure.

### `stream.end()`

Calling `stream.end()` will close the WebSocket with the code `1000` and reason `'stream end'`. `1000` indicates a normal closure, meaning that the purpose for which the connection was established has been fulfilled. (https://tools.ietf.org/html/rfc6455#section-7.4.1)

Clients may implement this to mean that the server is closing the stream intentionally, and the client should not automatically reconnect.

```javascript
const stream = new WebSocketJSONStream(ws)
// Closes WebSocket with the code 1000 and the reason 'stream end'
stream.end()
```

The code `1000` may also be used when calling the [`webSocket.close(code)`](https://html.spec.whatwg.org/multipage/web-sockets.html#dom-websocket-close) method of WebSockets in browsers.

### `stream.destroy()`

Calling `stream.destroy()` without an error object will close the stream without a code. This results in the client emitting a CloseEvent that has code `1005` and reason `''`. `1005` is a reserved value and MUST NOT be set as a status code in a Close control frame by an endpoint. It is designated for use in applications expecting a status code to indicate that no status code was actually present. (https://tools.ietf.org/html/rfc6455#section-7.4.1)

```javascript
const stream = new WebSocketJSONStream(ws)
// Closes WebSocket with the code 1005 and the reason ''
stream.destroy()
```

Calling `webSocket.close()` method of WebSockets in browsers without any arguments will produce a CloseEvent with the code `1005`. A reason string cannot be provided together with the code `1005`.

### `stream.destroy(error)`

Calling `stream.destroy(error)` with an error will emit an `'error'` event and close the stream with the code `1011` and reason `'stream error'` by default. `1011` indicates that a remote endpoint is terminating the connection because it encountered an unexpected condition that prevented it from fulfilling the request. (http://www.rfc-editor.org/errata_search.php?eid=3227)

```javascript
const stream = new WebSocketJSONStream(ws)
stream.on('error', (error) => {
  // Error event must be handled, or it will be throw when calling
  // stream.destroy() with an error argument
})

// Closes WebSocket with the code 1011 and the reason 'stream error'
const error = new Error('Unexpected server error')
stream.destroy(error)
```

The code `1011` cannot be used when calling the `webSocket.close(code)` method of WebSockets in browsers.

### `error.closeCode` and `error.closeReason`

Custom close code and reason values may be sent by setting `error.closeCode` or `error.closeReason` properties on the error argument passed to `stream.destroy(error)`. For example:

```javascript
const stream = new WebSocketJSONStream(ws)
stream.on('error', (error) => {
  // Error event must be handled, or it will be throw when calling
  // stream.destroy() with an error argument
})

// Example of extending from Error and adding additional properties
class CustomStreamError extends Error {
    constructor(message) {
        super(message)
        this.name = this.constructor.name
        Error.captureStackTrace(this, this.constructor)
        this.closeCode = null
        this.closeReason = null
    }
}

// Closes WebSocket with the code 4000 and the reason 'custom reason'.
// error.message is not sent to the client
const error = new CustomStreamError('Example error')
error.closeCode = 4000
error.closeReason = 'custom reason'
stream.destroy(error)
```

Browser WebSockets allow custom close codes between 3000 and 4999.


CHANGELOG.md
# 1.1.0

## Non-Breaking Changes

- A correct WebSocket "close" reason is now reported based on how the stream is ended/destroyed.


# 1.0.0

Initial release.


test.js
const assert = require('chai').assert
const http = require('http')
const WebSocket = require('ws')
const WebSocketJSONStream = require('./index')

const handler = (done, code) => (...args) => {
    try {
        code(...args)
        done()
    } catch (e) {
        done(e)
    }
}

describe('WebSocketJSONStream', function () {
    beforeEach(function (done) {
        this.httpServer = http.createServer()
        this.wsServer = new WebSocket.Server({ server: this.httpServer })
        this.httpServer.listen(() => {
            const address = this.httpServer.address()

            this.url = `http://${address.address}:${address.port}`
            this.connect(({ clientWebSocket, serverWebSocket }) => {
                this.clientStream = new WebSocketJSONStream(this.clientWebSocket = clientWebSocket)
                this.serverStream = new WebSocketJSONStream(this.serverWebSocket = serverWebSocket)
                done()
            })
        })

        this.connect = function (callback) {
            const clientWebSocket = new WebSocket(this.url)

            this.wsServer.once('connection', serverWebSocket => {
                clientWebSocket.once('open', () =>
                    callback({ clientWebSocket, serverWebSocket }))
            })
        }
    })

    afterEach(function (done) {
        this.wsServer.close()
        this.httpServer.close(done)
    })

    it('should send and receive messages', function (done) {
        this.connect(({ clientWebSocket, serverWebSocket }) => {
            const serverSentData = [ { a: 1 }, { b: 2 } ]
            const clientSentData = [ { y: -1 }, { z: -2 } ]
            const serverReceivedData = []
            const clientReceivedData = []

            serverSentData.forEach(data => this.serverStream.write(data))
            clientSentData.forEach(data => this.clientStream.write(data))

            this.serverStream.on('data', data => serverReceivedData.push(data))
            this.clientStream.on('data', data => clientReceivedData.push(data))

            this.clientStream.end()
            this.clientStream.on('close', () => {
                try {
                    assert.deepEqual(serverReceivedData, clientSentData)
                    assert.deepEqual(clientReceivedData, serverSentData)
                    done()
                } catch (e) {
                    done(e)
                }
            })
        })
    })

    it('should get clientStream close on clientStream.end()', function (done) {
        this.clientStream.end()
        this.clientStream.on('close', () => done())
    })
    it('should get clientStream close on serverStream.end()', function (done) {
        this.serverStream.end()
        this.clientStream.on('close', () => done())
    })
    it('should get serverStream close on clientStream.end()', function (done) {
        this.clientStream.end()
        this.serverStream.on('close', () => done())
    })
    it('should get serverStream close on serverStream.end()', function (done) {
        this.serverStream.end()
        this.serverStream.on('close', () => done())
    })

    it('should get clientStream close on clientStream.destroy()', function (done) {
        this.clientStream.destroy()
        this.clientStream.on('close', () => done())
    })
    it('should get clientStream close on serverStream.destroy()', function (done) {
        this.serverStream.destroy()
        this.clientStream.on('close', () => done())
    })
    it('should get serverStream close on clientStream.destroy()', function (done) {
        this.clientStream.destroy()
        this.serverStream.on('close', () => done())
    })
    it('should get serverStream close on serverStream.destroy()', function (done) {
        this.serverStream.destroy()
        this.serverStream.on('close', () => done())
    })

    it('should get clientStream finish on clientStream.end()', function (done) {
        this.clientStream.end()
        this.clientStream.on('finish', () => done())
    })
    it('should get serverStream finish on serverStream.end()', function (done) {
        this.serverStream.end()
        this.serverStream.on('finish', () => done())
    })
    it('should get serverStream end on clientStream.end()', function (done) {
        this.clientStream.end()
        this.serverStream.resume()
        this.serverStream.on('end', () => done())
    })
    it('should get clientStream end on serverStream.end()', function (done) {
        this.serverStream.end()
        this.clientStream.resume()
        this.clientStream.on('end', () => done())
    })

    it('should get clientStream error on clientWebSocket error', function (done) {
        const error = new Error('test')
        this.clientStream.once('error', handler(done, e => assert.strictEqual(e, error)))
        this.clientWebSocket.emit('error', error)
    })
    it('should get serverStream error on serverWebSocket error', function (done) {
        const error = new Error('test')
        this.clientStream.once('error', handler(done, e => assert.strictEqual(e, error)))
        this.clientWebSocket.emit('error', error)
    })
    it('should get clientStream error on clientStream.write invalid data (function)', function (done) {
        this.clientStream.once('error', handler(done, e => assert.instanceOf(e, Error)))
        this.clientStream.write(function () {})
    })
    it('should get serverStream error on serverStream.write invalid data (function)', function (done) {
        this.serverStream.once('error', handler(done, e => assert.instanceOf(e, Error)))
        this.serverStream.write(function () {})
    })
    it('should get clientStream error on clientStream.write invalid data (cyclic data)', function (done) {
        const data = {}
        data.a = data
        this.clientStream.once('error', handler(done, e => assert.instanceOf(e, Error)))
        this.clientStream.write(data)
    })
    it('should get serverStream error on serverStream.write invalid data (cyclic data)', function (done) {
        const data = {}
        data.a = data
        this.serverStream.once('error', handler(done, e => assert.instanceOf(e, Error)))
        this.serverStream.write(data)
    })
    it('should get clientStream error on serverWebSocket.send invalid data', function (done) {
        this.clientStream.once('error', handler(done, e => assert.instanceOf(e, Error)))
        this.serverWebSocket.send('qwerty')
    })
    it('should get serverStream error on clientWebSocket.send invalid data', function (done) {
        this.serverStream.once('error', handler(done, e => assert.instanceOf(e, Error)))
        this.clientWebSocket.send('qwerty')
    })
    it('should get clientStream error on clientStream.write after end', function (done) {
        this.clientStream.once('error', handler(done, e => assert.instanceOf(e, Error)))
        this.clientStream.end()
        this.clientStream.write({})
    })
    it('should get serverStream error on serverStream.write after end', function (done) {
        this.serverStream.once('error', handler(done, e => assert.instanceOf(e, Error)))
        this.serverStream.end()
        this.serverStream.write({})
    })
    it('should get clientStream error on clientStream.write, if clientWebSocket is closed', function (done) {
        this.clientWebSocket.close()
        this.clientWebSocket.on('close', () => {
            this.clientStream.once('error', handler(done, e => assert.instanceOf(e, Error)))
            this.clientStream.write({})
        })
    })
    it('should get serverStream error on serverStream.write, if serverWebSocket is closed', function (done) {
        this.serverWebSocket.close()
        this.serverWebSocket.on('close', () => {
            this.serverStream.once('error', handler(done, e => assert.instanceOf(e, Error)))
            this.serverStream.write({})
        })
    })
    it('should get clientStream error when clientWebSocket is not open yet', function (done) {
        const clientWebSocket = new WebSocket(this.url)
        const clientStream = new WebSocketJSONStream(clientWebSocket)
        let e = null

        // Delay `done` to avoid afterEach failure.
        clientWebSocket.on('open', handler(done, () => assert.instanceOf(e, Error)))
        clientStream.on('error', error => e = error)
        clientStream.write({})
    })
    it('should get clientStream error when clientWebSocket sends JSON-encoded null', function (done) {
        this.clientStream.on('error', handler(done, e => assert.instanceOf(e, Error)))
        this.serverWebSocket.send('null')
    })
    it('should get clientStream error when clientWebSocket sends JSON-encoded undefined', function (done) {
        this.clientStream.on('error', handler(done, e => assert.instanceOf(e, Error)))
        this.serverWebSocket.send('undefined')
    })

    it('should not crash on clientStream.destroy when clientWebSocket.readyState === WebSocket.CONNECTING', function (done) {
        const clientWebSocket = new WebSocket(this.url)
        const clientStream = new WebSocketJSONStream(clientWebSocket)

        clientStream.on('close', () => done())
        clientStream.destroy()
    })
    it('should not crash on clientStream.destroy when clientWebSocket.readyState === WebSocket.CONNECTING and gets error', function (done) {
        const clientWebSocket = new WebSocket('http://invalid-url:0')
        const clientStream = new WebSocketJSONStream(clientWebSocket)

        clientStream.on('error', () => null) // ignore invalid-url error
        clientStream.on('close', () => done())
        clientStream.destroy()
    })
    it('should not crash on clientStream.destroy when clientWebSocket.readyState === WebSocket.CLOSED', function (done) {
        const clientWebSocket = new WebSocket(this.url)

        clientWebSocket.on('open', () => {
            clientWebSocket.close()
            clientWebSocket.on('close', () => {
                const clientStream = new WebSocketJSONStream(clientWebSocket)

                clientStream.destroy()
                done()
            })
        })
    })
})


test/close-status.js
const assert = require('chai').assert

const NORMAL_CLOSURE_CODE = 1000
const NO_STATUS_CODE = 1005
const INTERNAL_ERROR_CODE = 1011
const CUSTOM_CODE = 4000

module.exports = function() {
    ['clientWebSocket', 'serverWebSocket'].forEach(webSocketName => {
        describe(webSocketName + ' CloseEvent', function () {
            it('is emitted with code 1000 and reason "stream end" on clientStream.end()', function (done) {
                this[webSocketName].addEventListener('close', event => {
                    assert.equal(event.code, NORMAL_CLOSURE_CODE)
                    assert.equal(event.reason, 'stream end')
                    done()
                })
                this.clientStream.end()
            })
            it('is emitted with code 1000 and reason "stream end" on serverStream.end()', function (done) {
                this[webSocketName].addEventListener('close', event => {
                    assert.equal(event.code, NORMAL_CLOSURE_CODE)
                    assert.equal(event.reason, 'stream end')
                    done()
                })
                this.serverStream.end()
            })
            it('is emitted with code 1005 and reason "" on clientStream.destroy()', function (done) {
                this[webSocketName].addEventListener('close', event => {
                    assert.equal(event.code, NO_STATUS_CODE)
                    assert.equal(event.reason, '')
                    done()
                })
                this.clientStream.destroy()
            })
            it('is emitted with code 1005 and reason "" on serverStream.destroy()', function (done) {
                this[webSocketName].addEventListener('close', event => {
                    assert.equal(event.code, NO_STATUS_CODE)
                    assert.equal(event.reason, '')
                    done()
                })
                this.serverStream.destroy()
            })
            it('is emitted with code 1011 and reason "stream error" on clientStream.destroy(error)', function (done) {
                this[webSocketName].addEventListener('close', event => {
                    assert.equal(event.code, INTERNAL_ERROR_CODE)
                    assert.equal(event.reason, 'stream error')
                    done()
                })
                const error = new Error('Test error')
                this.clientStream.on('error', () => {})
                this.clientStream.destroy(error)
            })
            it('is emitted with code 1011 and reason "stream error" on serverStream.destroy(error)', function (done) {
                this[webSocketName].addEventListener('close', event => {
                    assert.equal(event.code, INTERNAL_ERROR_CODE)
                    assert.equal(event.reason, 'stream error')
                    done()
                })
                const error = new Error('Test error')
                this.serverStream.on('error', () => {})
                this.serverStream.destroy(error)
            })
            it('is emitted with code from error.closeCode on clientStream.destroy(error)', function (done) {
                this[webSocketName].addEventListener('close', event => {
                    assert.equal(event.code, CUSTOM_CODE)
                    done()
                })
                const error = new Error('Test error')
                error.closeCode = CUSTOM_CODE
                this.clientStream.on('error', () => {})
                this.clientStream.destroy(error)
            })
            it('is emitted with code from error.closeCode on serverStream.destroy(error)', function (done) {
                this[webSocketName].addEventListener('close', event => {
                    assert.equal(event.code, CUSTOM_CODE)
                    done()
                })
                const error = new Error('Test error')
                error.closeCode = CUSTOM_CODE
                this.serverStream.on('error', () => {})
                this.serverStream.destroy(error)
            })
            it('is emitted with reason from error.closeReason on clientStream.destroy(error)', function (done) {
                this[webSocketName].addEventListener('close', event => {
                    assert.equal(event.reason, 'custom reason')
                    done()
                })
                const error = new Error('Test error')
                error.closeReason = 'custom reason'
                this.clientStream.on('error', () => {})
                this.clientStream.destroy(error)
            })
            it('is emitted with reason from error.closeReason on serverStream.destroy(error)', function (done) {
                this[webSocketName].addEventListener('close', event => {
                    assert.equal(event.reason, 'custom reason')
                    done()
                })
                const error = new Error('Test error')
                error.closeReason = 'custom reason'
                this.serverStream.on('error', () => {})
                this.serverStream.destroy(error)
            })
        })

        describe(webSocketName + ' "close" event', function () {
            it('is emitted with code 1000 and reason "stream end" on clientStream.end()', function (done) {
                this[webSocketName].on('close', (code, reason) => {
                    assert.equal(code, NORMAL_CLOSURE_CODE)
                    assert.equal(reason, 'stream end')
                    done()
                })
                this.clientStream.end()
            })
            it('is emitted with code 1000 and reason "stream end" on serverStream.end()', function (done) {
                this[webSocketName].on('close', (code, reason) => {
                    assert.equal(code, NORMAL_CLOSURE_CODE)
                    assert.equal(reason, 'stream end')
                    done()
                })
                this.serverStream.end()
            })
            it('is emitted with code 1005 and reason "" on clientStream.destroy()', function (done) {
                this[webSocketName].on('close', (code, reason) => {
                    assert.equal(code, NO_STATUS_CODE)
                    assert.equal(reason, '')
                    done()
                })
                this.clientStream.destroy()
            })
            it('is emitted with code 1005 and reason "" on serverStream.destroy()', function (done) {
                this[webSocketName].on('close', (code, reason) => {
                    assert.equal(code, NO_STATUS_CODE)
                    assert.equal(reason, '')
                    done()
                })
                this.serverStream.destroy()
            })
            it('is emitted with code 1011 and reason "stream error" on clientStream.destroy(error)', function (done) {
                this[webSocketName].on('close', (code, reason) => {
                    assert.equal(code, INTERNAL_ERROR_CODE)
                    assert.equal(reason, 'stream error')
                    done()
                })
                const error = new Error('Test error')
                this.clientStream.on('error', () => {})
                this.clientStream.destroy(error)
            })
            it('is emitted with code 1011 and reason "stream error" on serverStream.destroy(error)', function (done) {
                this[webSocketName].on('close', (code, reason) => {
                    assert.equal(code, INTERNAL_ERROR_CODE)
                    assert.equal(reason, 'stream error')
                    done()
                })
                const error = new Error('Test error')
                this.serverStream.on('error', () => {})
                this.serverStream.destroy(error)
            })
            it('is emitted with code from error.closeCode on clientStream.destroy(error)', function (done) {
                this[webSocketName].on('close', (code) => {
                    assert.equal(code, CUSTOM_CODE)
                    done()
                })
                const error = new Error('Test error')
                error.closeCode = CUSTOM_CODE
                this.clientStream.on('error', () => {})
                this.clientStream.destroy(error)
            })
            it('is emitted with code from error.closeCode on serverStream.destroy(error)', function (done) {
                this[webSocketName].on('close', (code) => {
                    assert.equal(code, CUSTOM_CODE)
                    done()
                })
                const error = new Error('Test error')
                error.closeCode = CUSTOM_CODE
                this.serverStream.on('error', () => {})
                this.serverStream.destroy(error)
            })
            it('is emitted with reason from error.closeReason on clientStream.destroy(error)', function (done) {
                this[webSocketName].on('close', (code, reason) => {
                    assert.equal(reason, 'custom reason')
                    done()
                })
                const error = new Error('Test error')
                error.closeReason = 'custom reason'
                this.clientStream.on('error', () => {})
                this.clientStream.destroy(error)
            })
            it('is emitted with reason from error.closeReason on serverStream.destroy(error)', function (done) {
                this[webSocketName].on('close', (code, reason) => {
                    assert.equal(reason, 'custom reason')
                    done()
                })
                const error = new Error('Test error')
                error.closeReason = 'custom reason'
                this.serverStream.on('error', () => {})
                this.serverStream.destroy(error)
            })
        })
    })
}


destroy.js
// Copied from:
// https://github.com/nodejs/node/blob/f5a2167eb5565b84e068b385dc22c5e74f543d1b/lib/internal/streams/destroy.js

'use strict';

/* istanbul ignore next */
// undocumented cb() API, needed for core, not for public API
function destroy(err, cb) {
  const readableDestroyed = this._readableState &&
    this._readableState.destroyed;
  const writableDestroyed = this._writableState &&
    this._writableState.destroyed;

  if (readableDestroyed || writableDestroyed) {
    if (cb) {
      cb(err);
    } else if (err &&
               (!this._writableState || !this._writableState.errorEmitted)) {
      process.nextTick(emitErrorNT, this, err);
    }
    return this;
  }

  // we set destroyed to true before firing error callbacks in order
  // to make it re-entrance safe in case destroy() is called within callbacks

  if (this._readableState) {
    this._readableState.destroyed = true;
  }

  // if this is a duplex stream mark the writable part as destroyed as well
  if (this._writableState) {
    this._writableState.destroyed = true;
  }

  this._destroy(err || null, (err) => {
    if (!cb && err) {
      process.nextTick(emitErrorAndCloseNT, this, err);
      if (this._writableState) {
        this._writableState.errorEmitted = true;
      }
    } else if (cb) {
      process.nextTick(emitCloseNT, this);
      cb(err);
    } else {
      process.nextTick(emitCloseNT, this);
    }
  });

  return this;
}

/* istanbul ignore next */
function emitErrorAndCloseNT(self, err) {
  emitErrorNT(self, err);
  emitCloseNT(self);
}

/* istanbul ignore next */
function emitCloseNT(self) {
  if (self._writableState && !self._writableState.emitClose)
    return;
  if (self._readableState && !self._readableState.emitClose)
    return;
  self.emit('close');
}

/* istanbul ignore next */
function emitErrorNT(self, err) {
  self.emit('error', err);
}

module.exports = {
  destroy
};


index.js
const Duplex = require('stream').Duplex
const WebSocket = require('ws')
const destroy = require('./destroy').destroy
const noop = function () {}

const NORMAL_CLOSURE_CODE = 1000
const NORMAL_CLOSURE_REASON = 'stream end'
const INTERNAL_ERROR_CODE = 1011
const INTERNAL_ERROR_REASON = 'stream error'

module.exports = class WebSocketJSONStream extends Duplex {
    constructor(ws) {
        super({
            objectMode: true,
            allowHalfOpen: false,
            emitClose: false
        })

        this.ws = ws;

        this.ws.on('message', message => {
            let value

            try {
                value = JSON.parse(message)
            } catch (error) {
                return this.emit('error', error)
            }

            if (value == null) {
                return this.emit('error', new Error('Can\'t JSON.parse the value'))
            }

            this.push(value)
        })

        this.ws.on('close', () => {
            this.push(null)
            this.emit('close')
        })

        this.ws.on('error', error => {
            this.emit('error', error)
        })

        // Required by nodejs 6.X.X which does not support `_final`.
        this.once('finish', () => this._closeOnStreamEnd(noop))

        // Required by nodejs 6.X.X which does not support `destroy`.
        /* istanbul ignore if */
        if (typeof this.destroy !== 'function') {
            this.destroy = destroy
        }
    }

    _read() {}

    _write(object, encoding, callback) {
        let json

        try {
            json = JSON.stringify(object)
        } catch (error) {
            return callback(error)
        }

        if (typeof json !== 'string') {
            return callback(new Error('Can\'t JSON.stringify the value'))
        }

        this.ws.send(json, callback)
    }

    _final(callback) {
        this._closeOnStreamEnd(callback)
    }

    _destroy(error, callback) {

        /*
         * Calling destroy without an error object will close the stream
         * without a code. This results in the client emitting a CloseEvent
         * that has code 1005.
         *
         * 1005 is a reserved value and MUST NOT be set as a status code in a
         * Close control frame by an endpoint.  It is designated for use in
         * applications expecting a status code to indicate that no status
         * code was actually present.
         * https://tools.ietf.org/html/rfc6455#section-7.4.1
         */
        let code
        let reason

        if (error) {

            /*
             * 1011 indicates that a remote endpoint is terminating the
             * connection because it encountered an unexpected condition that
             * prevented it from fulfilling the request.
             * http://www.rfc-editor.org/errata_search.php?eid=3227
             */
            code = error.closeCode || INTERNAL_ERROR_CODE
            reason = error.closeReason || INTERNAL_ERROR_REASON
        }
        this._closeWebSocket(code, reason, () => callback(error))
    }

    _closeOnStreamEnd(callback) {

        /*
         * 1000 indicates a normal closure, meaning that the purpose for which
         * the connection was established has been fulfilled.
         * https://tools.ietf.org/html/rfc6455#section-7.4.1
         */
        this._closeWebSocket(NORMAL_CLOSURE_CODE, NORMAL_CLOSURE_REASON, callback)
    }

    _closeWebSocket(code, reason, callback) {
        switch (this.ws.readyState) {
            case WebSocket.CONNECTING:
                this.ws.once('error', () => this._closeWebSocket(code, reason, callback))
                this.ws.once('open', () => this._closeWebSocket(code, reason, callback))
                break
            case WebSocket.OPEN:
                this.ws.once('close', () => callback())
                this.ws.close(code, reason)
                break
            case WebSocket.CLOSING:
                this.ws.once('close', () => callback())
                break
            case WebSocket.CLOSED:
                process.nextTick(callback)
                break
            /* istanbul ignore next */
            default:
                process.nextTick(callback, new Error(`Unexpected readyState: ${this.ws.readyState}`))
                break
        }
    }
}
